src/player.cpp

```diff
@@ -33,22 +34,101 @@ void Player::setStateChangedCallback(void (*stateChangedCallback)(unsigned int))
     soloud.setStateChangedCallback(stateChangedCallback);
 }
 
-PlayerErrors Player::init(unsigned int sampleRate, unsigned int bufferSize, unsigned int channels)
+PlayerErrors Player::init(unsigned int sampleRate, unsigned int bufferSize, unsigned int channels, int deviceID)
 {
     if (mInited)
         return playerAlreadyInited;
 
+    void *playbackInfos_id = nullptr;
+    if (deviceID != -1)
+    {
+        // Calling this will init [pPlaybackInfos]
+        auto const devices = listPlaybackDevices();
+        if (devices.size() == 0 || deviceID >= devices.size())
+            return backendNotInited;
+        playbackInfos_id = &pPlaybackInfos[deviceID].id;
+    }
+
     // initialize SoLoud.
     SoLoud::result result = soloud.init(
         SoLoud::Soloud::CLIP_ROUNDOFF,
-        SoLoud::Soloud::MINIAUDIO, sampleRate, bufferSize, channels);
-    if (result == SoLoud::SO_NO_ERROR)
+        SoLoud::Soloud::MINIAUDIO, sampleRate, bufferSize, channels, playbackInfos_id);
+    if (result == SoLoud::SO_NO_ERROR){
         mInited = true;
-    else
+        mSampleRate = sampleRate;
+        mBufferSize = bufferSize;
+        mChannels = channels;
+    } else
         result = backendNotInited;
     return (PlayerErrors)result;
 }
 
+PlayerErrors Player::changeDevice(int deviceID)
+{
+    if (!mInited)
+        return backendNotInited;
+
+    void *playbackInfos_id = nullptr;
+
+    // Calling this will init [pPlaybackInfos]
+    auto const devices = listPlaybackDevices();
+    if (devices.size() == 0 || deviceID >= devices.size())
+        return backendNotInited;
+    playbackInfos_id = &pPlaybackInfos[deviceID].id;
+
+    SoLoud::result result = soloud.miniaudio_changeDevice(playbackInfos_id);
+    
+    if (result != SoLoud::SO_NO_ERROR)
+        result = backendNotInited;
+    return noError;
+}
+
+// List available playback devices.
+std::vector<PlaybackDevice> Player::listPlaybackDevices()
+{
+    // printf("***************** LIST DEVICES START\n");
+    ma_context context;
+    ma_uint32 playbackCount;
+    ma_device_info *pCaptureInfos;
+    ma_uint32 captureCount;
+    std::vector<PlaybackDevice> ret;
+    ma_result result;
+    if ((result = ma_context_init(NULL, 0, NULL, &context)) != MA_SUCCESS)
+    {
+        printf("Failed to initialize context %d\n", result);
```

Comment:
I'm not sure I'm a big fan of just printing to console and returning an empty vector but I guess this deep down it's just a nit.

Reply:
Replaced with just a comment. If an empty array is returned I suppose is because the hardware doesn't have a playback device.